### DISCUSSION TITLE
Help with functional interface

### DISCUSSION THREAD

**Discussion by nik2606** [magpylib/magpylib#790] [Q&A] [answered]
Hi magpylib community,

I think I have a problem understanding the usage of the functional interface, hope someone can help me with this. My goal is to model the B-Field of a multiple magnet configuration. I can easily create the magnet configuration in the object-oriented way, to test if my code can be sped up, I wanted to try out the functional way. 

Take the following code as an example. Lets say I want to create 100 observer points and 10 cuboid magnets:
```python
import numpy as np
import magpylib as magpy

B = magpy.getB(
    sources='Cuboid',
    observers=meshgrid_aoi, #Just a np.array of xyz-coordinates, shape 100,3
    dimension=np.full((10,3), 12),
    position=np.linspace((0,0,0), (0,0,10), 10),
    polarization=np.full((10,3), 12),
)

print(B)
```
When running this code, I get the following error:

> MagpylibBadUserInput: Input array lengths must be 1 or of a similar length.
> Instead received lengths {'dimension': 10, 'polarization': 10, 'observers': 100, 'position': 10}

It is obviously not possible to create a configuration where number of magnets is not equal to number of observers. So is the functional interface not designed for problems involving multiple magnet inputs? Is there another way to handle my magnet/observer configuration?

Thanks for your help!

**Comment by Alexboiboi** [comment, marked as answer]
Hi @nik2606

Thanks for your interest in the Magpylib library!

The error you're encountering arises because the `getB` function (in functional interface mode) expects input arrays  to have lengths of either 1 or to match the number of observers. This is necessary for the function to correctly pair the sources and observers. Essentially, you need to broadcast your magnet properties across all observer points (as the object oriented interface does) an the output will have the same length as each input.

Here's a modified version of your code to accomplish this:

```python
import numpy as np
import magpylib as magpy

# Define number of magnets and observers
num_magnets = 10
num_observers = 100

# Create observer points (meshgrid_aoi) - shape (100, 3)
meshgrid_aoi = np.random.rand(num_observers, 3)

# Create magnet properties
dimensions = np.full((num_magnets, 3), 12)        # shape (10, 3)
positions = np.linspace((0, 0, 0), (0, 0, 10), num_magnets)  # shape (10, 3)
polarizations = np.full((num_magnets, 3), 12)     # shape (10, 3)

# Repeat each magnet property for each observer and magnet
dimensions_expanded = np.tile(dimensions, (num_observers, 1))
positions_expanded = np.tile(positions, (num_observers, 1))
polarizations_expanded = np.tile(polarizations, (num_observers, 1))
observers_expanded = np.repeat(meshgrid_aoi, num_magnets, axis=0)

# Now call the functional getB function
B = magpy.getB(
    sources='Cuboid',
    observers=observers_expanded,
    dimension=dimensions_expanded,
    position=positions_expanded,
    polarization=polarizations_expanded,
)
```

Hope this helps,

Alex -

**Reply by nik2606** [reply]
Hi Alex, got it! Thanks for your explanation!

**Comment by OrtnerMichael** [comment]
comment: The functional interface requires length n for all inputs,  or of length 1 which will be repeated to length n automatically, for computing n instances.

The documentation or the docstrings were not helpful ? Can we improve there ?

**Reply by nik2606** [reply]
Hard to tell. My mistake probably was that I constantly kept overlooking the fact that the functional interface is basically calculating parameter sets for every source instead of calculating the whole magnet/observer constellation at once. With Alexes explanation/code this now makes sense to me. Maybe you could add another example to your documentation, similar to the one above. 
That being said I am not a trained computer scientist, so my lack of experience/knowlege probably also contributed to this.

Thank you so much for the work you all put into magpylib, it helped me a lot!

**Reply by OrtnerMichael** [reply]
thank you for clarifying - magpylib is not for comuputer scientists but for physicists and engineers.

we will try to make this more clear, also by separating the functional interface more clearly from the object oriented one.
